Support querying PCI address

cathay4t opened the pull request:
By reading link of `/sys/class/net/<nic_name>/device`, we include
`PciAddress` to `Iface`.

Example output of `npc iface enp7s0`:

```
pci_address: 0000:07:00.0
```

Example output of `npc enp7s0`:

```
3: enp7s0: <UP,BROADCAST,RUNNING,MULTICAST,LOWERUP> state up mtu 1500
  driver e1000e pci 0000:07:00.0
  link ethernet
  mac 00:bb:cc:dd:ee:ff permanent_mac 00:bb:cc:dd:ee:ff
```

No test case included as this require PCI device which does exists in CI
environment.

cathay4t commented:
The PCI address only available via ethtool IOCTL `ETHTOOL_GDRVINFO`, there is no netlink API for this yet. Hence using `sysfs` required.

The systemd `src/udev/udev-builtin-path_id.c` is also using this information for `ID_PATH`.

tyll commented:
netdevsim virtual devices have a PCI address. Not sure, if it works as when I checkout your branch, compile it with `make` and run `./target/debug/npc eth0` I don't see the PCI address.

tyll commented on src/cli/npc.rs:
```diff
@@ -193,7 +200,7 @@ impl CliIfaceBrief {
                         if let Some(fwd) = ip_info.forwarding {
-                            addr_strs.push(format!("forwarding {}", fwd));
+                            addr_strs.push(format!("forwarding {fwd}"));
                         }
```
This seems to be a unrelated change.

tyll replied:
There seem to be several other unrelated changes in this commit like this.

cathay4t replied:
The second commit is for cargo clippy and cargo fmt.

ihuguet replied:
The first commit also contains lot of cargo clippy/fmt changes, making the review more difficult. In following PRs please try to split and structure in logical and small commits to facilitate reviews.

tyll commented on src/lib/query/pci.rs:
```diff
@@ -0,0 +1,111 @@
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut items: Vec<&str> = s.split(":").collect();
+        let err = NisporError::invalid_argument(format!(
```
For USB interfaces the PCI address is something like `2-2.4.4.3:1.0`. I don't see this error for a USB or netdevsim interface (which uses `netdevsim1` as PCI address).

cathay4t replied:
I do not have plan to support USB interface.

tyll commented:
> netdevsim virtual devices have a PCI address. Not sure, if it works as when I checkout your branch, compile it with `make` and run `./target/debug/npc eth0` I don't see the PCI address.

This is because eth0 is a  USB device here and the patch does not support this. The PCI address is shown for the wifi card, though.

cathay4t commented:
netdevsim does not have PCI address.

ihuguet commented on src/lib/query/pci.rs:
```diff
@@ -0,0 +1,111 @@
+
+impl std::fmt::Display for PciAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{:02x}:{:02x}.{}",
+            if self.domain > u16::MAX.into() {
+                format!("{:08x}", self.domain)
+            } else {
+                format!("{:04x}", self.domain)
+            },
+            self.bus,
+            self.slot,
+            self.function
+        )
```
Better
```
       write!(
            f,
            if self.domain > u16::MAX.into() {
                "{:08x}:{:02x}:{:02x}.{}"
            } else {
                 "{:04x}:{:02x}:{:02x}.{}"
            },
            self.domain,
            self.bus,
            self.slot,
            self.function
        )
```
It avoids a string allocation.

cathay4t replied:
```
error: format argument must be a string literal
  --> src/lib/query/pci.rs:24:13
   |
24 | /             if self.domain > u16::MAX.into() {
25 | |                 "{:08x}:{:02x}:{:02x}.{}"
26 | |             } else {
27 | |                 "{:04x}:{:02x}:{:02x}.{}"
28 | |             },
   | |_____________^
   |
help: you might be missing a string literal to format with
   |
24 |             "{} {} {} {} {}", if self.domain > u16::MAX.into() {
   |             +++++++++++++++++

error: could not compile `nispor` (lib) due to 1 previous error
```

ihuguet replied:
You're right.

ihuguet commented on src/lib/query/pci.rs:
```diff
@@ -0,0 +1,111 @@
+            "Invalid PCI address, expecting string \
+                like 00:1b.0 or 0000:07:00.1, but got {s}"
+        ));
+        match items.len() {
+            3 | 2 => {
+                // It is safe to unwrap as we already checked the length.
+                let slot_function_str: Vec<&str> =
+                    items.pop().unwrap().split(".").collect();
+                if slot_function_str.len() != 2 {
+                    return Err(err.clone());
+                }
+
+                let slot = u8::from_str_radix(slot_function_str[0], 16)
+                    .map_err(|_| err.clone())?;
```
You can avoid creating the error beforehand and doing err.clone() all the time. Make the `match` to return an `Option<PciAddress>`. Then, if you got `None`, create the error and return it.
Now it always create/allocate the error once. If there is a real error it does a clone.
In the other way it doesn't do anything if there is no error, and it creates the error once if there is an error.

cathay4t replied:
Created a function `fn parse_pci_addr(pci_addr_str: &str) -> Option<PciAddress>` allowing ? to work there.

ihuguet commented on src/lib/query/pci.rs:
```diff
@@ -0,0 +1,111 @@
+}
+
+impl Iface {
+    pub(crate) fn fill_pci_address(&mut self) {
+        if let Some(dev_path) =
+            PathBuf::from(format!("/sys/class/net/{}/device", self.name))
+                .read_link()
```
Can you add a comment like https://github.com/nispor/nispor/blob/071a2c8c0e406b2248a3082961df7efa057d8aad/src/lib/query/iface.rs#L594?
It will be easier to find it and remember when the netlink interface becomes available.

cathay4t replied:
Added.

cathay4t committed:
Run cargo fmt and cargo clippy

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
Support querying PCI address

By reading link of `/sys/class/net/<nic_name>/device`, we
include `PciAddress` to `Iface`.

Example output of `npc iface enp7s0`:

```
pci_address: 0000:07:00.0
```

Example output of `npc enp7s0`:

```
 3: enp7s0: <UP,BROADCAST,RUNNING,MULTICAST,LOWERUP> state up mtu 1500
    driver e1000e pci 0000:07:00.0
    link ethernet
    mac 00:bb:cc:dd:ee:ff permanent_mac 00:bb:cc:dd:ee:ff
```

No test case included as this require PCI device which does exists in CI
environment.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
New patch uploaded. Changes:
 1. Introduce `parse_pci_addr()` returning `Option<PciAddress>` for `?` to work without cloning the error.
 2. Placed all lint changes to dedicate commit.
 3. Unit test case included.

ihuguet commented:
LGTM

tyll commented:
> netdevsim does not have PCI address.

Here it does and Ansible reports it also, for example:

```
ansible  -m setup localhost  | grep pciid
            "pciid": "netdevsim1",
            "pciid": "netdevsim1",
            "pciid": "2-2.4.4.3:1.0",
            "pciid": "0000:09:00.0",
```
